pkg/network/ebpf/c/prebuilt/tracer.c

```diff
@@ -592,16 +642,58 @@ int kprobe__tcp_sendmsg__pre_4_1_0(struct pt_regs* ctx) {
     return handle_message(&t, size, 0);
 }
 
+static __always_inline void http_prepare_key(u32 cpu, http_batch_key_t *key, http_batch_state_t *batch_state) {
+    __builtin_memset(key, 0, sizeof(http_batch_key_t));
+    key->cpu = cpu;
+    key->page_num = batch_state->idx % HTTP_BATCH_PAGES;
+}
+
+static __always_inline void http_notify_batch(struct pt_regs* ctx) {
+    u32 cpu = bpf_get_smp_processor_id();
+
+    http_batch_state_t *batch_state = bpf_map_lookup_elem(&http_batch_state, &cpu);
+    if (batch_state == NULL || batch_state->pos < HTTP_BATCH_SIZE) {
+        return;
+    }
+
+    http_batch_key_t key;
+    http_prepare_key(cpu, &key, batch_state);
+
+    http_batch_t *batch = bpf_map_lookup_elem(&http_batches, &key);
+    if (batch == NULL) {
+        return;
+    }
+
+    // It's important to zero the struct so we account for the padding
+    // introduced by the compilation, otherwise you get a `invalid indirect read
+    // from stack off`. Alternatively we can either use a #pragma pack directive
+    // or try to manually add the padding to the struct definition. More
+    // information in https://docs.cilium.io/en/v1.8/bpf/ under the
+    // alignment/padding section
+    http_batch_notification_t notification = {0};
+    notification.cpu = cpu;
+    notification.batch_idx = batch_state->idx;
+
+    bpf_perf_event_output(ctx, &http_notifications, cpu, &notification, sizeof(http_batch_notification_t));
+    log_debug("http batch notification flushed: cpu: %d idx: %d lost_events: %d\n", cpu, batch->idx, batch_state->pos-HTTP_BATCH_SIZE);
+    batch_state->idx++;
+    batch_state->pos = 0;
+}
+
 SEC("kretprobe/tcp_sendmsg")
 int kretprobe__tcp_sendmsg(struct pt_regs* ctx) {
+#if DEBUG == 1
     int ret = PT_REGS_RC(ctx);
 
     log_debug("kretprobe/tcp_sendmsg: return: %d\n", ret);
+
     // If ret < 0 it means an error occurred but we still counted the bytes as being sent
     // let's increment our miscount count
     if (ret < 0) {
         increment_telemetry_count(tcp_sent_miscounts);
```

Comment:
Why is this being put into a debug-only block?

Reply:
this `kretprobe` was being used for debug purposes before. now that this is being shared with the http program, I wrapped it in debug block to minimize the overhead when http monitoring is on but debug is off.